### PR TITLE
[16.0][MIG] - sale_stock_picking_blocking

### PR DESCRIPTION
**Pull request by sbejaoui**

### CODE REVIEW

**Comment by sbejaoui** [comment]
#2215

File: sale_stock_picking_blocking/models/sale_order.py [hunk 1]
```diff
@@ -26,9 +26,9 @@ def _check_not_auto_done(self):
             )
 
     @api.onchange("partner_id")
-    def onchange_partner_id(self):
+    def _onchange_partner_id_warning(self):
         """Add the 'Default Delivery Block Reason' if set in the partner."""
-        res = super().onchange_partner_id()
+        res = super()._onchange_partner_id_warning()
```
Comment (lmignon) [inline sale_stock_picking_blocking/models/sale_order.py]: @sbejaoui  `delivery_block_id` should become a compted field with `readonly=False`. It is important that no more business logic depends on onchange methods to make sure that the logic is applied even if the creation is done by a call to the create method from within the code.

**Review by lmignon — CHANGES_REQUESTED** [review]

**Review by lmignon — APPROVED** [review]
LGTM (Code review)

**Comment by lmignon** [comment]
LGTM (Functional tests)

File: sale_stock_picking_blocking/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    @api.constrains("delivery_block_id")
+    def _check_not_auto_done(self):
+        auto_done = self.user_has_groups("sale.group_auto_done_setting")
+        if auto_done and any(so.delivery_block_id for so in self):
+            raise ValidationError(
+                _('You cannot block a sale order with "auto_done_setting" ' "active.")
+            )
+
+    @api.onchange("partner_id")
```
Comment (rousseldenis) [inline sale_stock_picking_blocking/models/sale_order.py]:
```suggestion
    @api.depends("partner_id")
```

File: sale_stock_picking_blocking/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        """Add the 'Default Delivery Block Reason' if set in the partner."""
+        for so in self:
+            so.delivery_block_id = so.partner_id.default_delivery_block or False
+
+    def action_remove_delivery_block(self):
+        """Remove the delivery block and create procurements as usual."""
+        for order in self.filtered(
+            lambda so: so.state == "sale" or not so.delivery_block_id
+        ):
+            order.write({"delivery_block_id": False})
```
Comment (rousseldenis) [inline sale_stock_picking_blocking/models/sale_order.py]:
```suggestion
            orders_to_unblock = self.filtered(...)
            orders_to_unblock.write({"delivery_block_id": False})
            orders_to_unblock.mapped("order_line").(...)
```

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

**Review by rousseldenis — APPROVED** [review]
Code review

**Pull request merged**